partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rel/RelExclude.kt

```diff
@@ -0,0 +1,184 @@
+            val newRecord = exclusions.fold(row) { curRecord, expr ->
+                excludeOnRecord(curRecord, expr)
+            }
+            return newRecord
+        }
+    }
+
+    override fun close() {
+        input.close()
+    }
+
+    @OptIn(PartiQLValueExperimental::class)
+    private fun excludeOnRecord(
+        record: Record,
+        exclusions: CompiledExcludeItem
+    ): Record {
+        val values = record.values
+        val value = values.getOrNull(exclusions.root)
+        val newValues = if (value != null) {
+            values[exclusions.root] = excludeOnPartiQLValue(value, exclusions)
+            values
+        } else {
+            values
+        }
+        return Record(newValues)
+    }
+
+    @OptIn(PartiQLValueExperimental::class)
+    private fun excludeOnStructValue(
+        structValue: StructValue<*>,
+        exclusions: ExcludeNode
+    ): PartiQLValue {
+        val attrsToRemove = exclusions.leaves.mapNotNull { leaf ->
+            when (val leafStep = leaf.step) {
+                is ExcludeStep.StructWildcard -> {
+                    // tuple wildcard at current level. return empty struct
+                    return structValue<PartiQLValue>()
+                }
+                is ExcludeStep.StructField -> leafStep.attr
+                is ExcludeStep.CollIndex, is ExcludeStep.CollWildcard -> null
+            }
+        }.toSet()
+        val branches = exclusions.branches
+        val finalStruct = structValue.entries.mapNotNull { structField ->
+            if (attrsToRemove.contains(structField.first)) {
+                // struct attr is to be removed at current level
+                null
+            } else {
+                // deeper level exclusions
+                val name = structField.first
+                var expr = structField.second
+                // apply case-sensitive tuple attr exclusions at deeper levels
+                val structFieldCaseSensitiveKey = ExcludeStep.StructField(name, ExcludeFieldCase.SENSITIVE)
+                branches.find {
+                    it.step == structFieldCaseSensitiveKey
+                }?.let {
+                    expr = excludeOnPartiQLValue(expr, it)
+                }
+                // apply case-insensitive tuple attr exclusions at deeper levels
+                val structFieldCaseInsensitiveKey = ExcludeStep.StructField(name, ExcludeFieldCase.INSENSITIVE)
+                branches.find {
+                    it.step == structFieldCaseInsensitiveKey
+                }?.let {
+                    expr = excludeOnPartiQLValue(expr, it)
+                }
+                // apply tuple wildcard exclusions at deeper levels
+                val tupleWildcardKey = ExcludeStep.StructWildcard
+                branches.find {
+                    it.step == tupleWildcardKey
+                }?.let {
+                    expr = excludeOnPartiQLValue(expr, it)
+                }
+                Pair(name, expr)
```

Comment:
You're looping over the branches multiple times by using find. A set has O(1) lookup when you are using the hash function for the set, but in this case you are using `find` which is a loop over the set elements.

You should only have one loop of the branches. Does the order in which exclusions apply matter? If so, consider "pruning" early such applying wildcards before single fields as to eliminate unnecessary work. That being said, the planner should be responsible for reducing the exclusions.

Reply:
Changed to store the lookups in a map. The order of exclusions should not matter. In latest revision, planner step now eliminates the redundant exclude paths.